lightning/src/ln/peers/handler.rs

```diff
@@ -527,220 +465,374 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 	///
 	/// Panics if the descriptor was not previously registered in a new_*_connection event.
 	pub fn read_event(&self, peer_descriptor: &mut Descriptor, data: &[u8]) -> Result<bool, PeerHandleError> {
-		match self.do_read_event(peer_descriptor, data) {
-			Ok(res) => Ok(res),
-			Err(e) => {
-				self.disconnect_event_internal(peer_descriptor, e.no_connection_possible);
-				Err(e)
-			}
-		}
+		self.inner.read_event(peer_descriptor, data)
 	}
 
-	/// Append a message to a peer's pending outbound/write buffer, and update the map of peers needing sends accordingly.
-	fn enqueue_message<M: Encode + Writeable>(&self, peers_needing_send: &mut HashSet<Descriptor>, peer: &mut Peer, descriptor: Descriptor, message: &M) {
-		let mut buffer = VecWriter(Vec::new());
-		wire::write(message, &mut buffer).unwrap(); // crash if the write failed
-		let encoded_message = buffer.0;
-
-		log_trace!(self.logger, "Enqueueing message of type {} to {}", message.type_id(), log_pubkey!(peer.their_node_id.unwrap()));
-		match peer.encryptor {
-			PeerState::Connected(ref mut conduit) => peer.pending_outbound_buffer.push_back(conduit.encrypt(&encoded_message[..])),
-			_ => panic!("peer must be connected!")
-		}
-		peers_needing_send.insert(descriptor);
+	/// Checks for any events generated by our handlers and processes them. Includes sending most
+	/// response messages as well as messages generated by calls to handler functions directly (eg
+	/// functions like ChannelManager::process_pending_htlc_forward or send_payment).
+	pub fn process_events(&self) {
+		self.inner.process_events();
 	}
 
-	fn do_read_event(&self, peer_descriptor: &mut Descriptor, data: &[u8]) -> Result<bool, PeerHandleError> {
-		let pause_read = {
-			let mut peers_lock = self.peers.lock().unwrap();
-			let peers = &mut *peers_lock;
-			let pause_read = match peers.peers.get_mut(peer_descriptor) {
-				None => panic!("Descriptor for read_event is not already known to PeerManager"),
-				Some(peer) => {
-
-					let mut send_init_message = false;
+	/// Indicates that the given socket descriptor's connection is now closed.
+	///
+	/// This must only be called if the socket has been disconnected by the peer or your own
+	/// decision to disconnect it and must NOT be called in any case where other parts of this
+	/// library (eg PeerHandleError, explicit disconnect_socket calls) instruct you to disconnect
+	/// the peer.
```

Comment:
Naming is confusing, I've to read twice before to remember that `disconnect_socket` and `socket_disconnected` aren't  the same thing. Maybe lay out  events order :
- timer_tick_occurred()
- disconnect_socket()
- socket_disconnect()

Reply:
100% agree.   Still trip on myself remembering which one is an event callback and which is the action.

What about introducing a suffix like `socket_disconnected_cb` to make it clear that it is called due to an event? If you like this, what is the process to get feedback on public API changes?

Reply:
I think a good name would be `on_socket_disconnected` which should hint better the callback notion. At least that's how we name one of our callbacks in `keysinterface.rs. We don't have a strict naming policy for callback so best advice I can give you. Thus if you have better prefix/suffix let's go for it.

Pre-0.1, we don't have a formalized API change process so just make a clear commit about it.

Reply:
Tracked here: https://github.com/rust-bitcoin/rust-lightning/issues/712